### PR TITLE
* Replaced Ktor with Scarlet for ease of use

### PR DESCRIPTION
**Pull request by TalhaAli00**
* Created the WalletConnectClient (mirror of Client in JS codebase)
* Added the pairing function to WalletConnectClient that accepts a PairParams object containing the Dapp URI
* Added classes to represent the Pairing Proposal, Pairing Approval, and Pairing Reject as per the Out-of-Bound section in the Tech Spec
* Happy path test done for parsing a URI into a Pairing Proposal object
* Sample test included but commented out to show how Scarlet Websocket Client and MockWebServer will communicate during unit tests
* Updated Kotlin to 1.5.30

### CODE REVIEW

File: walletconnectV2/src/main/kotlin/org/walletconnect/walletconnectv2/DefaultWebSocketClient.kt [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+//    suspend fun wsConnect(block: suspend (DefaultClientWebSocketSession.() -> Unit) = {}) {
+//        client.webSocket(urlString = getURL(), block = block)
+//    }
+
+    fun test() {
+        relay.send("".toByteArray())
+    }
+
+    private fun getServerUrl(): String = if (useTLs) "wss://$hostName:$port" else "ws://$hostName:$port"
```
Comment (ligi) [inline walletconnectV2/src/main/kotlin/org/walletconnect/walletconnectv2/DefaultWebSocketClient.kt:41]:
```suggestion
    private fun getServerUrl(): String = if (useTLs) "wss://$hostName:$port" else "ws://$hostName:$port"
    private fun getServerUrl(): String = (if (useTLs) "wss" else "ws")+"://$hostName:$port"
```
shorter and easier to spot where the difference is

**Review by ligi — CHANGES_REQUESTED** [review]
Some nits/suggestions after a shallow review.
Also IMHO we should lower the usage of (companion) objects - sure it's easy but it smells. And can be often just replaced with top level constants/functions. And there is stuff that is just needed for tests leaking into the main codebase (the initLocal thing - can't see use-cases in the lib for consumers for that stuff)

File: sample/build.gradle.kts [hunk 2]
```diff
@@ -32,5 +32,9 @@ android {
 }
 
 dependencies {
+    implementation("androidx.core:core-ktx:1.3.2")
```
[the remark this answers is not in the document]
Comment (TalhaAli00) [reply sample/build.gradle.kts:35]: Default dependencies when creating an android project. I'll review the dependencies once the sample is built and remove this if I don't end up using anything from the dependency.

**Comment by TalhaAli00** [comment]
> Some nits/suggestions after a shallow review.
> Also IMHO we should lower the usage of (companion) objects - sure it's easy but it smells. And can be often just replaced with top level constants/functions. And there is stuff that is just needed for tests leaking into the main codebase (the initLocal thing - can't see use-cases in the lib for consumers for that stuff)

@ligi 
The primary reason for containing everything in an object is so functions can be annotated with JvmStatic. One goal while developing the SDK is to have a seamless experience regardless of if the consumer of the SDK is working in a Java file or Kotlin file. The only other way of achieving this would be to have top level functions and then annotate the file with JvmName. I'm not sure if one way is better than the other. I could look into the bytecode java files and see how it looks when going through and optimizing

File: walletconnectV2/src/main/kotlin/org/walletconnect/walletconnectv2/DefaultWebSocketClient.kt [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    // job might be unnecessary since client has it's own coroutine context
+//    private val job = SupervisorJob()
+//    private val scope = CoroutineScope(job + Dispatchers.IO)
+
+//    suspend fun wsConnect(block: suspend (DefaultClientWebSocketSession.() -> Unit) = {}) {
+//        client.webSocket(urlString = getURL(), block = block)
+//    }
+
+    fun test() {
+        relay.send("".toByteArray())
```
[the remark this answers is not in the document]
Comment (TalhaAli00) [reply walletconnectV2/src/main/kotlin/org/walletconnect/walletconnectv2/DefaultWebSocketClient.kt:38]: done

File: walletconnectV2/src/main/kotlin/org/walletconnect/walletconnectv2/DefaultWebSocketClient.kt (hunk 1 above)
Comment (TalhaAli00) [reply walletconnectV2/src/main/kotlin/org/walletconnect/walletconnectv2/DefaultWebSocketClient.kt:41]: updated in current branch

**Pull request merged**